### PR TITLE
cli: new lifecycle CLI command "init"

### PR DESCRIPTION
**Pull request by cjdcordeiro**
Adds a new CLI command `init` for initializing a new Rockcraft project. It populates the current directory with a minimal, but yet functional, rockcraft.yaml file.

The content of this template rockcraft.yaml depends on the new properties defined in https://github.com/canonical/rockcraft/pull/34

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
-----

### CODE REVIEW

**Comment by cjdcordeiro** [comment]
ping @valentincanonical

File: rockcraft/lifecycle.py [hunk 1]
```diff
@@ -26,8 +27,38 @@
 from .parts import PartsLifecycle, Step
 from .project import Project, load_project
 from .providers import capture_logs_from_instance
+from rockcraft.errors import RockcraftInitError
 
 
+def init(rockcraft_yaml_content: str) -> None:
+    """Initialize a rockcraft project.
+
+    Args:
```
Comment (sergiusens) [inline rockcraft/lifecycle.py:36]: note that we use rst for our documentation

File: tests/unit/test_cli.py [hunk 2]
```diff
@@ -25,10 +25,18 @@
...
 def lifecycle_pack_mock():
-    """Mock for ui.init."""
+    """Mock for ui.pack."""
     patcher = patch("rockcraft.lifecycle.pack")
     yield patcher.start()
     patcher.stop()
+    
+
+@pytest.fixture
+def lifecycle_init_mock():
```
Comment (sergiusens) [inline tests/unit/test_cli.py:35]: you could use the mocker fixture for simplicity directly in the test method

**Comment by sergiusens** [comment]
I guess with the spawn of all projects we forgot about putting our preffered commit structure on the README.md here (https://github.com/canonical/craft-parts/#committing-code)

File: rockcraft/lifecycle.py (hunk 1 above)
Comment (cjdcordeiro) [reply rockcraft/lifecycle.py:36]: good catch...bad autodoc setup on the IDE. Fixed

File: tests/unit/test_cli.py (hunk 2 above)
Comment (cjdcordeiro) [reply tests/unit/test_cli.py:35]: agreed. That's my common practice as well. However, this seems to be the strategy in all tests. 

I'd suggest going through all of these in a separate refactoring thread: https://github.com/canonical/rockcraft/issues/37

**Comment by cjdcordeiro** [comment]
> I guess with the spawn of all projects we forgot about putting our preffered commit structure on the README.md here (https://github.com/canonical/craft-parts/#committing-code)

got it. Changing it for all my open PRs

File: rockcraft/errors.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+class ProviderError(CraftError):
+    """Error in provider operation."""
+
+
+class ProjectLoadError(CraftError):
+    """Error loading rockcraft.yaml."""
+
+
+class ProjectValidationError(CraftError):
+    """Error validatiing rockcraft.yaml."""
```
Comment (tigarmo) [inline rockcraft/errors.py:39]:
```suggestion
    """Error validating rockcraft.yaml."""
```

File: rockcraft/lifecycle.py [hunk 4]
```diff
@@ -18,17 +18,41 @@
...
 from .parts import PartsLifecycle, Step
 from .project import Project, load_project
 from .providers import capture_logs_from_instance
 
 
+def init(rockcraft_yaml_content: str) -> None:
+    """Initialize a rockcraft project.
+
+    :param rockcraft_yaml_content: Content of the rockcraft.yaml file
+    :type rockcraft_yaml_content: str
```
Comment (tigarmo) [inline rockcraft/lifecycle.py:39]: What's everyone's opinion regarding `:type` directives in docstrings for functions with type annotations? Personally I think we can skip these

File: rockcraft/lifecycle.py [hunk 5]
```diff
@@ -18,17 +18,41 @@
...
+    :raises RockcraftInitError: raises initialization error in case of conflicts
+    with existing rockcraft.yaml files
+    """
+    rockcraft_yaml_path = "rockcraft.yaml"
+
+    if os.path.exists(rockcraft_yaml_path):
```
Comment (tigarmo) [inline rockcraft/lifecycle.py:45]: Can/should we use `Path` instead?

File: rockcraft/lifecycle.py [hunk 6]
```diff
@@ -18,17 +18,41 @@
...
+        raise RockcraftInitError("{} already exists!".format(rockcraft_yaml_path))
+    elif os.path.exists(f".{rockcraft_yaml_path}"):
+        raise RockcraftInitError(f".{rockcraft_yaml_path} already exists!")
```
Comment (tigarmo) [inline rockcraft/lifecycle.py:48]: This is interesting, I didn't know rockcraft supported `.rockcraft.yaml`?

This one is minor but I thought it amusing that you chose `.format()` in some places and f-strings in other places, heh

File: tests/spread/general/init/task.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+summary: rockcraft init test
+
+execute: |
+  rockcraft init
+
```
Comment (tigarmo) [inline tests/spread/general/init/task.yaml:5]:
```suggestion
  test -f rockcraft.yaml
```

**Review by tigarmo — CHANGES_REQUESTED** [review]
thanks! I left some minor style suggestions, but the typo and the spread test I think are worth updating

File: rockcraft/errors.py (hunk 3 above)
Comment (cjdcordeiro) [reply rockcraft/errors.py:39]: uff :stuck_out_tongue: I still need to find a proper VSCode extension for grammar...one that doesn't complain about all the programming variables, but notices proper English mistakes

thanks

File: rockcraft/lifecycle.py (hunk 4 above)
Comment (cjdcordeiro) [reply rockcraft/lifecycle.py:39]: I agree. I suggest we do this in another PR though, cause `:type` is in quite a few docstrings.

https://github.com/canonical/rockcraft/issues/58

File: rockcraft/lifecycle.py (hunk 5 above)
Comment (cjdcordeiro) [reply rockcraft/lifecycle.py]: done

File: rockcraft/lifecycle.py (hunk 6 above)
Comment (cjdcordeiro) [reply rockcraft/lifecycle.py:46]: this is basically the "power" of the copy&paste https://github.com/snapcore/snapcraft/blob/main/snapcraft_legacy/internal/lifecycle/_init.py#L51

:rofl: 

fixed it

**Review by tigarmo — APPROVED** [review]
:+1:

File: rockcraft/commands/lifecycle.py [hunk 8]
```diff
@@ -46,3 +46,40 @@ class PackCommand(_LifecycleCommand):
...
+        """
+        Initialize a rockcraft project by creating a minimalist,
+        yet functional, rockcraft.yaml file in the current directory.
+        """
+    )
+
+    _INIT_TEMPLATE_YAML = textwrap.dedent(
+        """\
+            name: my-rock-name # the name of your ROCK
+            base: ubuntu:20.04 # the base environment for this ROCK
```
Comment (sergiusens) [inline rockcraft/commands/lifecycle.py:66]: can we not start out with 22.04?

File: rockcraft/errors.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Rockcraft error definitions."""
+
+from craft_cli import CraftError
```
Comment (sergiusens) [inline rockcraft/errors.py:19]: can you separate this refactor into a separate PR please?

File: rockcraft/commands/lifecycle.py [hunk 10]
```diff
@@ -46,3 +46,40 @@ class PackCommand(_LifecycleCommand):
...
+            version: '0.1' # just for humans. Semantic versioning is recommnended
+            summary: Single-line elevator pitch for your amazing ROCK # 79 char long summary
+            description: |
+                This is my my-rock-name's description. You have a paragraph or two to tell the
+                most important story about it. Keep it under 100 words though,
+                we live in tweetspace and your description wants to look good in the
+                container registries out there.
+            license: GPL-3.0 # your application's SPDX license
+
+            parts:
```
Comment (sergiusens) [inline rockcraft/commands/lifecycle.py:76]: with pebble being a default part we probably no longer need this, which is for the better!

Comment (cjdcordeiro) [reply rockcraft/commands/lifecycle.py:76]: no, you still need it.

`parts` is a mandatory field. The FieldValidation will fail if you don't pass it. Adding pebble as an implicit part happens after the YAML has been loaded. So `parts` must exist before that

File: rockcraft/commands/lifecycle.py (hunk 8 above)
Comment (cjdcordeiro) [reply rockcraft/commands/lifecycle.py]: y definitely (old commit)

done

File: rockcraft/errors.py (hunk 9 above)
Comment (cjdcordeiro) [reply rockcraft/errors.py]: sure: https://github.com/canonical/rockcraft/pull/61

File: rockcraft/commands/lifecycle.py (hunk 10 above)
Comment (sergiusens) [reply rockcraft/commands/lifecycle.py:76]: Yes, this can be done entry for pebble is added before we actually process the entry

**Comment by cmatsuoka** [comment]
Does init accept all the common lifecycle command line options? Otherwise it should be a separate, non-lifecycle command.

**Comment by tigarmo** [comment]
> 

I think most of the common options don't make sense for `init`, and conceptually I guess it's not really a part of the lifecycle (it's not a step, etc). I'll update.
